src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+    }
+
+    /// <summary>
+    /// Reports diagnostics on field declarations.
+    /// </summary>
+    /// <param name="context">The analysis context.</param>
+    /// <param name="classDeclarationSyntax">The class declaration node to find the fields in.</param>
+    /// <param name="fieldInitializationInfos">The initalization info for all fields present in a constructor.</param>
+    private static void ReportDiagnosticsOnFieldDeclarations(SyntaxNodeAnalysisContext context, ClassDeclarationSyntax classDeclarationSyntax, Dictionary<string, FieldInitializationInfo> fieldInitializationInfos)
+    {
+        foreach (FieldDeclarationSyntax field in classDeclarationSyntax.ChildNodes().OfType<FieldDeclarationSyntax>())
+        {
+            SeparatedSyntaxList<VariableDeclaratorSyntax> variables = field.Declaration.Variables;
+
+            if (variables.Count != 1)
+            {
+                // We only support single variable declarations for now.
+                continue;
+            }
+
+            VariableDeclaratorSyntax variable = variables[0];
+            EqualsValueClauseSyntax? initializer = variable.Initializer;
+            bool isInitializerPresent = initializer is not null;
+
+            if (!fieldInitializationInfos.TryGetValue(variable.Identifier.Text, out FieldInitializationInfo fieldInfo))
+            {
+                // Field was not found in constructors.
+                HandleFieldNotInConstructors(context, field, variable, initializer, isInitializerPresent);
+            }
+            else
+            {
+                // Field was found in constructors.
+                HandleFieldsInConstructors(context, field, initializer!, isInitializerPresent, fieldInfo);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Handles fields that are not initialized in constructors.
+    /// If the initializer is present, and the field is a value type that is initialied to null, zero, false or an empty struct,
+    /// a diagnostic is reported to remove the initialization in the declaration.
+    /// If the initializer is not present, and the field is not a value type or nullable, a diagnostic is reported to
+    /// initialize in declaration.
+    /// </summary>
+    /// <param name="context">The analysis context.</param>
+    /// <param name="field">The field declaration to analyze.</param>
+    /// <param name="variable">The variable declarator node.</param>
+    /// <param name="initializer">The field initializer.</param>
+    /// <param name="isInitializerPresent">A bool on whether the initializer already exists or not.</param>
+    private static void HandleFieldNotInConstructors(
+        SyntaxNodeAnalysisContext context,
+        FieldDeclarationSyntax field,
+        VariableDeclaratorSyntax variable,
+        EqualsValueClauseSyntax? initializer,
+        bool isInitializerPresent)
+    {
+        if (isInitializerPresent
+            && (IsFieldNullZeroOrFalseWithInitializer(context, initializer!)
```

Comment:
Don't use `!`

Reply:
Forgot to fix this in the refactor. Will address in the next iteration